xdsl/dialects/gpu.py

```diff
@@ -0,0 +1,124 @@
+            covariant=True)
+
+
+@irdl_attr_definition
+class _GPUAttr(ParametrizedAttribute, Generic[T]):
+    name = "gpu"
+
+    value: ParameterDef[T]
+
+    @staticmethod
+    def parse_parameters(parser: BaseParser) -> list[Attribute]:
+        parser.parse_characters("<", f"Expected <")
+        ntok = parser.tokenizer.next_token()
+
+        if ntok.text == "dim":
+            attrtype = _DimensionAttr
+            vtok = parser.tokenizer.next_token()
+            if vtok.text not in ["x", "y", "z"]:
+                parser.raise_error(
+                    f"Unexpected dim {vtok.text}. A gpu dim can only be x, y, or z",
+                    vtok)
+
+        elif ntok.text == "all_reduce_op":
+            attrtype = _AllReduceOperationAttr
+            vtok = parser.tokenizer.next_token()
+            if vtok.text not in [
+                    "add", "and", "max", "min", "mul", "or", "xor"
+            ]:
+                parser.raise_error(
+                    f"Unexpected op {vtok.text}. A gpu all_reduce_op can only be add, "
+                    "and, max, min, mul, or, or xor ", vtok)
+        else:
+            parser.raise_error(
+                f"Unexpected token {ntok.text}. Expected dim or all_reduce_op",
+                ntok)
+        parser.parse_characters(">", f"Expected >")
+        return [attrtype([StringAttr.from_str(vtok.text)])]
+
+    @classmethod
+    def from_op(cls: Type[_GPUAttr[_AllReduceOperationAttr]],
+                value: str) -> AllReduceOperationAttr:
+        return AllReduceOperationAttr(
+            [_AllReduceOperationAttr([StringAttr(value)])])
+
+    @property
+    def data(self) -> str:
+        return self.value.param.data
+
+    @classmethod
+    def from_dimension(cls: Type[_GPUAttr[_DimensionAttr]],
+                       value: str) -> DimensionAttr:
+        return DimensionAttr([_DimensionAttr([StringAttr(value)])])
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string("<")
+        self.value.print_parameters(printer)
+        printer.print_string(">")
+
+
+DimensionAttr = _GPUAttr[_DimensionAttr]
+AllReduceOperationAttr = _GPUAttr[_AllReduceOperationAttr]
+
+
+@irdl_op_definition
+class ModuleOp(Operation):
+    name = "gpu.module"
+
+    body: SingleBlockRegion
+    sym_name: OpAttr[StringAttr]
+
+    @staticmethod
+    def get(name: SymbolRefAttr, ops: list[Operation]) -> ModuleOp:
+        op = ModuleOp.build(attributes={"sym_name": name}, regions=[ops])
+        return op
+
+    def verify_(self):
+        if (len(self.body.ops) == 0
+                or not isinstance(self.body.ops[-1], ModuleEndOp)):
+            raise VerifyException("gpu.module must end with gpu.module_end")
+
+
+@irdl_op_definition
+class ModuleEndOp(Operation):
+    name = "gpu.module_end"
+
+    @staticmethod
+    def get() -> ModuleEndOp:
+        return ModuleEndOp.build()
+
+
+GPU = Dialect([ModuleOp, ModuleEndOp], [_GPUAttr])
```

Comment:
Shouldn't it be:

```suggestion
GPU = Dialect([ModuleOp, ModuleEndOp], [DimensionAttr, AllReduceOperationAttr])
```

I'm really not sure, @math-fehr might know more here.

Reply:
No, good catch, but this kinda is the hack!
xDSL is well-defined and thus constrained. To parse #gpu<>, I have to have a `gpu` named attribute, and attribute names have to be unique. So I have only one actual IRDL `gpu` attribute, which contains a dim or op.
What I could do is include them all in the parser, but then the dim one would be parsed e.g. `#dim dim x`, which is not a thing. So it wouldn't add anything more than a name clash risk

Reply:
Oooh, I get it. Man, that's pretty gross. Do we have an issue to get enum parsing into the parser somehow? Just so we can keep track of all the times where we should switch to that?

Reply:
This would be super helpful to write in a comment near the code!